atproto/auth/oauth/session.go

```diff
@@ -77,52 +80,42 @@ type ClientSession struct {
 	lk sync.RWMutex
 }
 
-// Requests new tokens from auth server, and returns the new access token on success.
-//
-// Internally takes a lock on session data around the entire refresh process, including retries. Persists data using [PersistSessionCallback] if configured.
-func (sess *ClientSession) RefreshTokens(ctx context.Context) (string, error) {
-	sess.lk.Lock()
-	defer sess.lk.Unlock()
-
-	body := RefreshTokenRequest{
-		ClientID:     sess.Config.ClientID,
-		GrantType:    "refresh_token",
-		RefreshToken: sess.Data.RefreshToken,
+// Helper method to handle DPoP retries and client assertions (if the client is confidential)
+// body object will be url-encoded (expected to be either RefreshTokenRequest or RevocationRequest)
+// expects sess.lk to be held by caller
+// if a non-nil *http.Response is returned, the caller is responsible for closing the response body
+func (sess *ClientSession) postToAuthServer(ctx context.Context, url string, body interface{}) (*http.Response, error) {
+	vals, err := query.Values(body)
+	if err != nil {
+		return nil, err
 	}
-	tokenURL := sess.Data.AuthServerTokenEndpoint
-
 	if sess.Config.IsConfidential() {
 		clientAssertion, err := sess.Config.NewClientAssertion(sess.Data.AuthServerURL)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
-		body.ClientAssertionType = &ClientAssertionJWTBearer
-		body.ClientAssertion = &clientAssertion
-	}
-
-	vals, err := query.Values(body)
-	if err != nil {
-		return "", err
+		vals.Set("client_assertion_type", ClientAssertionJWTBearer)
+		vals.Set("client_assertion", clientAssertion)
```

Comment:
Note: Because I'm setting the client assertion fields on the query string directly rather than using the struct fields, the client assertion struct fields aren't actually needed - but I figured they might be worth keeping around if we ever implement the server-side of oauth in go, too.